File: packages/server/src/workspaces/opencode-auth.ts
import crypto from "node:crypto"

export const OPENCODE_SERVER_USERNAME_ENV = "OPENCODE_SERVER_USERNAME" as const
export const OPENCODE_SERVER_PASSWORD_ENV = "OPENCODE_SERVER_PASSWORD" as const

export const DEFAULT_OPENCODE_USERNAME = "embeddedcowork" as const

export function generateOpencodeServerPassword(): string {
  return crypto.randomBytes(32).toString("base64url")
}

export function buildOpencodeBasicAuthHeader(params: { username?: string; password?: string }): string | undefined {
  const username = params.username
  const password = params.password

  if (!username || !password) {
    return undefined
  }

  const token = Buffer.from(`${username}:${password}`, "utf8").toString("base64")
  return `Basic ${token}`
}

export function resolveOpencodeServerAuth(options?: {
  userEnvironment?: Record<string, string | undefined>
}): { username: string; password: string; authorization: string } {
  const userEnv = options?.userEnvironment ?? {}

  const username = userEnv[OPENCODE_SERVER_USERNAME_ENV]
    ?? process.env[OPENCODE_SERVER_USERNAME_ENV]
    ?? DEFAULT_OPENCODE_USERNAME

  const password = userEnv[OPENCODE_SERVER_PASSWORD_ENV]
    ?? process.env[OPENCODE_SERVER_PASSWORD_ENV]
    ?? generateOpencodeServerPassword()

  const authorization = buildOpencodeBasicAuthHeader({ username, password })
  if (!authorization) {
    throw new Error("Failed to build OpenCode auth header")
  }

  return { username, password, authorization }
}

File: packages/server-rust/src/workspaces/opencode_auth.rs
use rand::Rng;

pub const OPENCODE_SERVER_USERNAME_ENV: &str = "OPENCODE_SERVER_USERNAME";
pub const OPENCODE_SERVER_PASSWORD_ENV: &str = "OPENCODE_SERVER_PASSWORD";
pub const DEFAULT_OPENCODE_USERNAME: &str = "embeddedcowork";

pub fn generate_opencode_server_password() -> String {
    let mut bytes = [0u8; 32];
    rand::thread_rng().fill(&mut bytes);
    use base64::Engine;
    base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(bytes)
}

pub fn build_opencode_basic_auth_header(username: &str, password: &str) -> Option<String> {
    if username.is_empty() || password.is_empty() {
        return None;
    }
    let token = base64_encode(&format!("{}:{}", username, password));
    Some(format!("Basic {}", token))
}

fn base64_encode(input: &str) -> String {
    use base64::Engine;
    base64::engine::general_purpose::STANDARD.encode(input)
}
